src/qibo/models/variational.py

```diff
@@ -292,3 +292,199 @@ def _loss(params, qaoa, hamiltonian):
+
+        Returns:
+            State vector after applying the QAOA exponential gates.
+        """
+        state = self.get_initial_state(initial_state)
+        self.calculate_callbacks(state)
+        n = int(self.params.shape[0])
+        for i in range(n // 2):
+            state = self._apply_exp(state, self.ham_solver,
+                                    self.params[2 * i])
+            state = self._apply_exp(state, self.mix_solver,
+                                    self.params[2 * i + 1])
+        return self.normalize_state(state)
+
+    def __call__(self, initial_state=None):
+        """Equivalent to :meth:`qibo.models.QAOA.execute`."""
+        return self.execute(initial_state)
+
+    def get_initial_state(self, state=None):
+        """"""
+        if self.accelerators is not None:
+            c = self.hamiltonian.circuit(self.params[0])
+            if state is None:
+                state = self.states.DistributedState.plus_state(c)
+            return c.get_initial_state(state)
+
+        if state is None:
+            return self.state_cls.plus_state(self.nqubits).tensor
+        return Circuit.get_initial_state(self, state)
+
+
+    def minimize(self, delta_t, max_layers, initial_state=None, tol=None, callback=None, compile=False, processes=None,):
+        """Optimizes the variational parameters of the FALQON.
+
+                Args:
+                    delta_t (float): initial guess for the time step. A too large delta_t will make the algorithm fail.
+                    max_layers (int): maximum number of layers allowed for the FALQON.
+                    initial_state (np.ndarray): initial state vector of the FALQON.
+                    tol (float): Tolerance of energy change. If not specified, no check is done.
+                    callback (callable): Called after each iteration for scipy optimizers.
+                    options (dict): a dictionary with options for the different optimizers.
+                    compile (bool): whether the TensorFlow graph should be compiled.
+                    processes (int): number of processes when using the paralle BFGS method.
+
+                Return:
+                    The final energy (expectation value of the ``hamiltonian``).
+                    The corresponding best parameters.
+                    None
+                """
+        from numpy import array, hstack, inf
+
+        parameters = array([delta_t, 0])
+
+        def _loss(params, falqon, hamiltonian):
+            falqon.set_parameters(params)
+            state = falqon(initial_state)
+            return hamiltonian.expectation(state)
+
+        energy = [inf]
+
+        if callback is not None:
+            callback_result = []
+
+        for it in range(1, max_layers + 1):
+            beta = _loss(parameters, self, self.evol_hamiltonian)
+
+            if tol is not None:
+                energy.append(_loss(parameters, self, self.hamiltonian).numpy())
```

Comment:
Prefer using `np.array` instead of `.numpy()`, because the latter will break if a numpy backend is used. I am aware that this problem exists in existing models (QAOA and VQE) so I will fix it in a different PR (thanks for helping in finding this!).